crates/biome_js_analyze/src/semantic_analyzers/nursery/use_jsx_key_in_iterable.rs

```diff
@@ -0,0 +1,295 @@
+            }
+        }
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            state,
+            markup! {
+                "Missing \"key\" prop for element in iterator"
+            },
+        )
+        .note(markup! {
+            "The order of the items may change, and having a key can help React identify which item was moved"
+        }).note(markup! {
+            "Check the "<Hyperlink href="https://react.dev/learn/rendering-lists#why-does-react-need-keys">"React documentation"</Hyperlink>". "
+        });
+        Some(diagnostic)
+    }
+}
+
+/// Handle collections of components
+///
+/// Examples
+///
+/// ```js
+/// [<h1></h1>, <h1></h1>]
+/// ```
+fn handle_collections(node: &JsArrayExpression, model: &SemanticModel) -> Vec<TextRange> {
+    node.elements()
+        .iter()
+        .filter_map(|node| {
+            let node = node.ok()?;
+            // no need to handle spread case, if the spread argument is itself a list it
+            // will be handled during list declaration
+            let node = node.as_any_js_expression()?;
+            let res = node
+                .syntax()
+                .descendants()
+                .filter_map(|x| x.cast::<ReactComponentExpression>())
+                .filter_map(|node| match node {
+                    ReactComponentExpression::JsxTagExpression(node) => handle_jsx_tag(&node),
+                    ReactComponentExpression::JsCallExpression(node) => {
+                        handle_react_non_jsx(&node, model)
+                    }
+                })
+                .collect::<Vec<_>>();
+            Some(res)
+        })
+        .flatten()
+        .collect()
+}
+
+/// Handle iterators return components
+///
+/// Examples
+///
+/// ```js
+/// data.map(x => <h1>{x}</h1>)
+/// ```
+fn handle_iterators(node: &JsCallExpression, model: &SemanticModel) -> Option<Vec<TextRange>> {
+    let callee = node.callee().ok()?;
+    let member_expression = AnyJsMemberExpression::cast_ref(callee.syntax())?;
+    let arguments = node.arguments().ok()?;
+
+    let mut callback_index = None;
+    let caller_name = member_expression
+        .object()
+        .ok()
+        .and_then(|o| o.as_js_identifier_expression()?.name().ok()?.name().ok());
+
+    if matches!(
+        member_expression.member_name()?.text(),
+        "map"
+            | "flatMap"
+            | "from"
+            | "forEach"
+            | "filter"
+            | "some"
+            | "every"
+            | "find"
+            | "findIndex"
+            | "reduce"
+            | "reduceRight"
+    ) {
+        if caller_name.is_some_and(|name| name == "Array") {
+            callback_index = Some(1);
+        } else {
+            callback_index = Some(0);
+        }
+    }
+
+    let callback_index = callback_index?;
```

Comment:
I suggest this refactoring:
```suggestion
    if !matches!(
        member_expression.member_name()?.text(),
        "map"
            | "flatMap"
            | "from"
            | "forEach"
            | "filter"
            | "some"
            | "every"
            | "find"
            | "findIndex"
            | "reduce"
            | "reduceRight"
    ) {
        return None;
    }
    let mut callback_index = None;
    let caller_name = member_expression
        .object()
        .ok()
        .and_then(|o| o.as_js_identifier_expression()?.name().ok()?.name().ok());
    let callback_index = if caller_name.is_some_and(|name| name == "Array") {
        1
    } else {
        0
    };
```